hydromt/data_catalog/drivers/raster/rasterio_driver.py

```diff
@@ -72,24 +75,42 @@ def read_data(
                 uris_cached.append(cached_uri)
             uris = uris_cached
 
-        # NoData part should be done in DataAdapter.
+        if mask is not None:
+            kwargs.update({"mosaic_kwargs": {"mask": mask}})
+
         if np.issubdtype(type(metadata.nodata), np.number):
             kwargs.update(nodata=metadata.nodata)
-        # TODO: Implement zoom levels in https://github.com/Deltares/hydromt/issues/875
-        # if zoom_level is not None and "{zoom_level}" not in uri:
-        #     zls_dict, crs = self._get_zoom_levels_and_crs(uris[0])
-        #     zoom_level = self._parse_zoom_level(
-        #         zoom_level, mask, zls_dict, crs
-        #     )
-        #     if isinstance(zoom_level, int) and zoom_level > 0:
-        #         # NOTE: overview levels start at zoom_level 1, see _get_zoom_levels_and_crs
-        #         kwargs.update(overview_level=zoom_level - 1)
 
-        if mask is not None:
-            kwargs.update({"mosaic_kwargs": {"mask": mask}})
+        # Fix overview level
+        if zoom:
+            try:
+                zls_dict: Dict[int, float] = metadata.zls_dict
+                crs: Optional[CRS] = metadata.crs
+            except AttributeError:  # pydantic extra=allow on SourceMetadata
+                zls_dict, crs = self.get_zoom_levels_and_crs(uris[0])
+
+            overview_level: Optional[int] = zoom_to_overview_level(
+                zoom, mask, zls_dict, crs
+            )
+            if overview_level:
+                # NOTE: overview levels start at zoom_level 1, see _get_zoom_levels_and_crs
+                kwargs.update(overview_level=overview_level - 1)
+
+        # If the metadata resolver has already resolved the overview level,
+        # trying to open zoom levels here will result in an error.
+        # Better would be to seperate uriresolver and driver: https://github.com/Deltares/hydromt/issues/1023
+        # Then we can implement looking for a overview level in the driver.
+        def _open() -> Union[xr.DataArray, xr.Dataset]:
+            try:
+                return open_mfraster(uris, **kwargs)
+            except rasterio.errors.RasterioIOError as e:
+                if "Cannot open overview level" in str(e):
+                    kwargs.pop("overview_level")
+                    return open_mfraster(uris, **kwargs)
+                else:
+                    raise
```

Comment:
should this be `raise e` ?

Reply:
[You can use both: https://stackoverflow.com/questions/36153805/difference-between-raise-and-raise-e]

Reply:
cool TIL